### ISSUE TITLE
Range doesn't accepts symbolic start, stop, step

### ISSUE THREAD

**Issue by czgdp1807** [sympy/sympy#16833]
#### Problem
Range doesn't accepts symbolic parameters like, `n`.
However, its analogous, `Interval` accepts such parameters.

#### Example of problem
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> n = symbols('n', positive=True)
>>> Range(0, n)
Traceback (most recent call last):
  File "/home/gagandeep/sympy/sympy/core/compatibility.py", line 419, in as_int
    raise TypeError
TypeError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gagandeep/sympy/sympy/sets/fancysets.py", line 453, in __new__
    for w in (start, stop, step)]
  File "/home/gagandeep/sympy/sympy/sets/fancysets.py", line 453, in <listcomp>
    for w in (start, stop, step)]
  File "/home/gagandeep/sympy/sympy/core/compatibility.py", line 425, in as_int
    raise ValueError('%s is not an integer' % (n,))
ValueError: n is not an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gagandeep/sympy/sympy/sets/fancysets.py", line 458, in __new__
    [0, 1/10, 1/5].'''))
ValueError: 
Finite arguments to Range must be integers; `imageset` can define
other cases, e.g. use `imageset(i, i/10, Range(3))` to give [0, 1/10,
1/5].

```

#### Other comments/references
I am working on making the `MultinomialDistribution` to accept symbolic parameters and [this line of code](https://github.com/sympy/sympy/blob/master/sympy/stats/joint_rv_types.py#L270) uses `n` which when becomes symbolic raises the exception as shown in the example on running the tests. It wouldn't be conceptually correct to use `Interval`. Hence, I think `Range` should be capable of handling symbolic parameters. 
PS - For the time being I am using `Interval`.

**Comment by Upabjojr** [comment]
`Range` should definitely accept symbolic intervals.

**Comment by czgdp1807** [comment]
Is following diff fine? I haven't tried the tests though and I am not aware of the consequences.
@smichr 
```diff
diff --git a/sympy/sets/fancysets.py b/sympy/sets/fancysets.py
index 7b556d049..5733a4b58 100644
--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -5,7 +5,7 @@
 from sympy.core.expr import Expr
 from sympy.core.function import Lambda
 from sympy.core.singleton import Singleton, S
-from sympy.core.symbol import Dummy, symbols
+from sympy.core.symbol import Dummy, symbols, Symbol
 from sympy.core.sympify import _sympify, sympify, converter
 from sympy.logic.boolalg import And
 from sympy.sets.sets import Set, Interval, Union, FiniteSet
@@ -441,11 +441,14 @@ def __new__(cls, *args):
 
         # expand range
         slc = slice(*args)
-
         if slc.step == 0:
             raise ValueError("step cannot be 0")
 
         start, stop, step = slc.start or 0, slc.stop, slc.step or 1
+
+        if any(isinstance(x, Symbol) for x in (start, stop, step)):
+            return Basic.__new__(cls, start, stop, step)
+
         try:
             start, stop, step = [
                 w if w in [S.NegativeInfinity, S.Infinity]
(END)
 from sympy.core.function import Lambda
 from sympy.core.singleton import Singleton, S
-from sympy.core.symbol import Dummy, symbols
+from sympy.core.symbol import Dummy, symbols, Symbol
 from sympy.core.sympify import _sympify, sympify, converter
 from sympy.logic.boolalg import And
 from sympy.sets.sets import Set, Interval, Union, FiniteSet
@@ -441,11 +441,14 @@ def __new__(cls, *args):
 
         # expand range
         slc = slice(*args)
-
         if slc.step == 0:
             raise ValueError("step cannot be 0")
 
         start, stop, step = slc.start or 0, slc.stop, slc.step or 1
+
+        if any(isinstance(x, Symbol) for x in (start, stop, step)):
+            return Basic.__new__(cls, start, stop, step)
+
         try:
             start, stop, step = [
                 w if w in [S.NegativeInfinity, S.Infinity]

```

**Comment by oscarbenjamin** [comment]
That will only work if start stop and step are symbols. Consider:
```
In [1]: isinstance(x+1, Symbol)                                                 
Out[1]: False
```

**Comment by oscarbenjamin** [comment]
Maybe testing `not x.is_number` would be what you want.

**Comment by czgdp1807** [comment]
@oscarbenjamin Any way to take such expressions into account. Should `Expr` be used instead of `Symbol`?

**Comment by czgdp1807** [comment]
Thanks @oscarbenjamin

**Comment by smichr** [comment]
Can start be pi? Can step be Float(0.1)?

**Comment by czgdp1807** [comment]
> Can start be pi? Can step be Float(0.1)?

I think yes. Currently, we can cannot pass, `Float` values too to `Range`. Sequences, like `[pi, 2*pi, 3*pi, . . . ]` cannot be generated as far as I have figured out using `Range`.

**Comment by asmeurer** [comment]
See also https://github.com/sympy/sympy/pull/11278. I don't know if any of that work was picked up. @smichr would know the latest status.

**Comment by asmeurer** [comment]
`Range` currently requires integers, the same as the built-in `range`. Maybe it would be better to treat it like `numpy.arange` instead?

**Comment by smichr** [comment]
I highly recommend against changes to Range unless there is something that it would allos that exsting Sets would not:
```python
>>> arange = imageset(i, a + b*i, S.Naturals0)
>>> a in arange
True
>>> a + b in arange
True
```

**Comment by asmeurer** [comment]
Maybe Range should be written to use imageset underneath the covers, if it already handles such things.

**Comment by Upabjojr** [comment]
I am still up for changing Range.

**Comment by asmeurer** [comment]
I don't know if `Range` should support non-integral steps, *a la* `arange`. I would leave that as an open question, until it is needed. But it absolutely should support symbolic parameters. The original point of `Range` was for use in code generation, to represent C-style for loops. We need to be able to represent loops with a symbolic number of steps. Using imageset is harder, because the code printers would need to know how to take an imageset and print it as a loop, which is not impossible obviously, but they already know how to print Range (`for(i=start;i<end;i+=step)`).

I do sympathize with @smichr that it could make the Range code more complicated than it already is (and it is already quite complicated, I remember the headaches of https://github.com/sympy/sympy/pull/10660 and https://github.com/sympy/sympy/pull/10723 quite well, actually, there's a reason I never got around to finishing https://github.com/sympy/sympy/pull/11278). So if we can simplify the code along the way, perhaps by deferring to something more general like imageset, then that would be great.

**Comment by smichr** [comment]
Would a stub `classs SymbolicRange(Range)` with no methods work for printing?

**Comment by asmeurer** [comment]
My thinking was to have the methods of Range give NotImplementedError for symbolic parameters, until the algorithms are implemented. Some of the algorithms would be complicated, and likely not needed for code generation or stats. I don't think a subclasses is needed for that. 

Or they could fallback to `self.rewrite(imageset).method`. That makes sense anyway, as not all operations on Range produce a range (for instance, `Range(0, oo, 2).union(Range(0, oo, 5))` is not a Range, but it could be represented by an imageset or conditionset).

**Comment by czgdp1807** [comment]
@asmeurer  I think `Range` should be kept as it is because @kangzhiq is working on something called `ArrayComprehension` in https://github.com/sympy/sympy/pull/16845 which will be more of like symbolic lists and hence, most of the work for symbolic dimensions can be accomplished by that. Moreover, I think adding a new class like, `SymbolicRange` would be a subset of `ArrayComprehension`. Let's see if there is very demanding issue which requires `Range` to accept symbols then we will think of that, otherwise, we should find out other methods to get things done.
PS - This issue will be open for getting comments and if we find any good requirement then we will reopen https://github.com/sympy/sympy/pull/16838 and resume the work.

**Comment by asmeurer** [comment]
I think `distribution.set` should return a Set object. Furthermore, are set operations like intersection or union used on the set objects? If so, you will have to have a real set, and support for those will need to be added to Range with symbolic parameters.

**Comment by czgdp1807** [comment]
@asmeurer I gave a thought again to symbolic ranges. So, basically `Range` is a set with elements at finite uniform spacing from each other. If we allow symbols then we have to make them uniform too. However, is it the right approach to have such symbolic sets? What if we do some changes to `imageset`? I am suggesting this because, `imageset` is much more dynamic, however it doesn't takes assumptions into account. Instead it tries to make a `Range` for discrete sets. Instead of making a `Range`, it should remain unevaluated and there should be a possibility of obtaining `inf` and `sup` out of those `imageset`. Will it be good?
If this doesn't sound good and symbolic `Range` is still a high priority then I think we can work pseudocode for such an idea as it will require considering many corner cases and complications. There is a detailed thread on the [above referenced]() PR.

**Comment by smichr** [comment]
Just as we remap certain imagesets to Range, we could have the reverse, too:
```
>> Range(a, b, c)
imageset(i, a + c*i, Intersection(Interval(0, ceiling((b - a)/c) - 1), S.Naturals0))
```

**Comment by czgdp1807** [comment]
Is it possible to obtain, `sup` and `inf` for such sets as atleast that's the most important information from any set one can obtain for computations?

**Comment by asmeurer** [comment]
Another thing that was brought up in another issue is that Range is ordered (and should stay so). So any representation of Range needs to keep the correct order. `Range(2, -1, -1)` and `Range(3)` are equal as sets but unequal as objects.

**Comment by smichr** [comment]
> My thinking was to have the methods of Range give NotImplementedError for symbolic parameters

Maybe we just add a flag `is_literal` or `is_symbolic` so each method knows what to do -- or not to do -- with the Range information. e.g. `Range(1, a, 3)` would only be displayed like that (and printed accordingly) but requesting `inf` would give a Piecewise result with an `a > 1` part and a `True` part.

**Comment by asmeurer** [comment]
If it simplifies things we could require that symbolic parameters are implicitly finite (just like they would be implicitly integral).

**Comment by czgdp1807** [comment]
Please read the thread https://github.com/sympy/sympy/pull/16962#discussion_r296674807. May be there is one more use case of symbolic ranges which can simplify things. Actually, there are two options either intersection should support `sup` and `inf` or we can make `Range` symbolic. What all do you say?

**Comment by oscarbenjamin** [comment]
I think it definitely should be possible to have `Range(n)` for a symbol `n`. It shouldn't be necessary to emulate Range using intersections or imagesets or anything: rather those things should simplify to Range when possible.

I don't mind whether Range supports non-integer arguments but it should be possible to use integer symbols and to substitute those symbols for integers.

**Comment by czgdp1807** [comment]
Yes, I think we can move step by step, first we can allow `Range` to have integer symbols, then if more use case come into picture like `pi` or other things we can allow `Range` accept them. A step by step approach would be better rather than thinking of all the use case in one go.

**Comment by Upabjojr** [comment]
Notice that the recently introduced `ArrayComprehension` can easily emulate `Range` with symbolic ranges, although the result is not a set (rather an array).

Anyway, +1 for adding symbolic support to `Range`. For those who do not like it, use NumPy!!!

**Comment by czgdp1807** [comment]
@smichr Is it possible to re-start the work on symbolic ranges.

**Comment by czgdp1807** [comment]
As, we have observed a lot of use cases of symbolic `Range` and many members support this feature, so I think I would work according to the following plan,

- First, I will focus on allowing only integer symbols, that is symbols which have the assumption, `integer=True` to `Range` and make changes for them only. We have observed a lot of use cases for this.

- Second, if more use cases found for scenarios like, `Range(pi, 6 * pi, pi)` then some work will be started in that direction. 

If no raised an objection for 24 hours then I will open a PR for the first goal of the plan. Thanks.
